hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieInstantTimeGenerator.java

```diff
@@ -52,36 +67,65 @@ public static String createNewInstantTime(long milliseconds) {
       String newCommitTime;
       do {
         Date d = new Date(System.currentTimeMillis() + milliseconds);
-        newCommitTime = INSTANT_TIME_FORMATTER.format(convertDateToTemporalAccessor(d));
+        newCommitTime = MILLIS_INSTANT_TIME_FORMATTER.format(convertDateToTemporalAccessor(d));
       } while (HoodieTimeline.compareTimestamps(newCommitTime, HoodieActiveTimeline.LESSER_THAN_OR_EQUALS, oldVal));
       return newCommitTime;
     });
   }
 
-  public static Date parseInstantTime(String timestamp) throws ParseException {
+  public static Date parseDateFromInstantTime(String timestamp) throws ParseException {
     try {
-      LocalDateTime dt = LocalDateTime.parse(timestamp, INSTANT_TIME_FORMATTER);
+      // Enables backwards compatibility with non-millisecond granularity instants
+      String timestampInMillis = timestamp;
+      if (isSecondGranularity(timestamp)) {
+        // Add milliseconds to the instant in order to parse successfully
+        timestampInMillis = timestamp + DEFAULT_MILLIS_EXT;
+      } else if (timestamp.length() > MILLIS_INSTANT_TIMESTAMP_FORMAT_LENGTH) {
+        // compaction and cleaning in metadata has special format. handling it by trimming extra chars and treating it with secs granularity
+        timestampInMillis = timestamp.substring(0, MILLIS_INSTANT_TIMESTAMP_FORMAT_LENGTH);
+      }
+
+      LocalDateTime dt = LocalDateTime.parse(timestampInMillis, MILLIS_INSTANT_TIME_FORMATTER);
       return Date.from(dt.atZone(ZoneId.systemDefault()).toInstant());
     } catch (DateTimeParseException e) {
       // Special handling for all zero timestamp which is not parsable by DateTimeFormatter
       if (timestamp.equals(ALL_ZERO_TIMESTAMP)) {
         return new Date(0);
       }
-      // compaction and cleaning in metadata has special format. handling it by trimming extra chars and treating it with secs granularity
-      if (timestamp.length() > INSTANT_TIMESTAMP_FORMAT_LENGTH) {
-        LocalDateTime dt = LocalDateTime.parse(timestamp.substring(0, INSTANT_TIMESTAMP_FORMAT_LENGTH), INSTANT_TIME_FORMATTER);
-        return Date.from(dt.atZone(ZoneId.systemDefault()).toInstant());
-      }
       throw e;
     }
   }
 
-  public static String formatInstantTime(Instant timestamp) {
-    return INSTANT_TIME_FORMATTER.format(timestamp);
+  private static boolean isSecondGranularity(String instant) {
+    return instant.length() == SECS_INSTANT_ID_LENGTH;
+  }
+
+  public static String getInstantForDate(Instant timestamp) {
```

Comment:
i just felt `getInstantForXXX()` confuses people as it sounds like returning java Instant. I think `formatInstantTime()`, `formatDate()` suit better, at least people expect it to return String.